### PR TITLE
Add brew cask outdated command

### PR DESCRIPTION
**Pull request by joshka**
Addresses the first part of https://github.com/caskroom/homebrew-cask/issues/29301
Does not (yet) have tests

I'm submitting this PR as a WIP for review. The absence of tests is intentional right now. I'll add tests before it gets merged.

-----

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

### CODE REVIEW

File: Library/Homebrew/cask/lib/hbc/cli/outdated.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      end
+
+      def self.help
+        "Show casks that have an updated version available"
+      end
+
+      def self.check_installed
+        casks = Hbc.installed
+        casks.each do |cask|
+          odebug "Checking outdated for Cask #{cask}"
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/outdated.rb]: Move this into the `check_outdated` method.

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 2]
```diff
@@ -85,6 +85,23 @@ def installed?
...
+        versions.each do |version|
+          if version.to_s.casecmp @latest_version
+            @latest_version=version.to_s
+          end
+        end
+      end
+      @latest_version
+    end
+
+    def outdated?(greedy)
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: `def outdated?(greedy: false)`

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 3]
```diff
@@ -85,6 +85,23 @@ def installed?
       !versions.empty?
     end
 
+    def latest_installed_version
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: `latest_installed_version` is actually just `versions.last`, since they are sorted by timestamp.

**Comment by joshka** [comment]
@reitermarkus done

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 4]
```diff
@@ -85,6 +85,15 @@ def installed?
...
+      versions.last
+    end
+
+    def outdated?(greedy: false)
+      (greedy || !auto_updates && !version == :latest) &&
+      !version.nil? && (version.casecmp latest_installed_version) == 1
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: I don't think `version` can be `nil`, so just check for `version == latest_installed_version`.

File: Library/Homebrew/cask/lib/hbc/cli/outdated.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+          rescue CaskUnavailableError => e
+            onoe e
+          end
+        end
+      end
+
+      def self.check_outdated(cask)
+        odebug "Checking outdated for Cask #{cask.token}"
+        if cask.outdated?(@greedy)
+          puts "#{cask.token}: (#{cask.latest_installed_version}) < #{cask.version}"
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/outdated.rb]:
```ruby
if $stdout.tty?
  puts "#{cask.token}: (#{cask.latest_installed_version}) < #{cask.version}"
else
  puts cask.token
end
```

File: Library/Homebrew/cask/lib/hbc/cli/outdated.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        cask_tokens = cask_tokens_from(args)
+        if cask_tokens.empty?
+          check_installed
+        else
+          check_casks(cask_tokens)
+        end
+      end
+
+      def self.help
+        "Show casks that have an updated version available"
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/outdated.rb]: Make `Casks` uppercase.

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 4 above)
Comment (vitorgalvao) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: A non-existing `version` would fail `audit`, if I’m not mistaken.

Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: Yes, and `version == latest_installed_version` would still work even if it was `nil`.

File: Library/Homebrew/cask/lib/hbc/cli/outdated.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      end
+
+      def self.check_installed
+        casks = Hbc.installed
+        casks.each do |cask|
+          check_outdated(cask)
+        end
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/outdated.rb]: `Hbc.installed.each(&method(:check_outdated))`

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 4 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: Version is nil when you've installed casks that aren't yet in a tap. E.g I have a couple of casks that I wrote that I submitted PRs for but never followed up to complete (sorry). They show up with a ! in brew cask list. This check is here as the call to casecmp fails otherwise. We need the casecmp for a later than check rather than an equality check.

Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: > We need the casecmp for a later than check rather than an equality check.

How is a case comparison helping with deciding whether a version is later than another given version? I think it's safe to assume that any other version other than the installed one is “later“, since I have never seen a Cask get downgraded.

So just make this `version != latest_version_installed`.

Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: I agree. More the point is outdated should mean "does not match the version in HBC"

Comment (vitorgalvao) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: > I have never seen a Cask get downgraded.

It has happened, but it’s mostly when the new version has a major bug that stops the app from running/installing, so even in those cases the downgrade can be considered an upgraded (you’d want the earlier version).

Either way, it has been discussed in the past we can’t trust any comparison other than “`version` is different” because our versions can be weird strings of characters. You can’t compare a hash to another and determine which was done later.

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 8]
```diff
@@ -85,6 +85,16 @@ def installed?
       !versions.empty?
     end
 
+    def latest_installed_version
+      versions.last
+    end
+
+    def outdated?(greedy)
+      (greedy || !auto_updates && !version == :latest) &&
+      !version.nil? &&  # ignore Casks without sources
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: I'd make this the first line, like this: `return false if version.nil?`

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 9]
```diff
@@ -85,6 +85,16 @@ def installed?
       !versions.empty?
     end
 
+    def latest_installed_version
+      versions.last
+    end
+
+    def outdated?(greedy)
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: `def outdated?(greedy: false)`

File: Library/Homebrew/cask/lib/hbc/cli/outdated.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            onoe e
+          end
+        end
+      end
+
+      def self.check_outdated(cask)
+        odebug "Checking outdated for Cask #{cask.token}"
+        if cask.outdated?(@greedy)
+          if $stdout.tty?
+            puts "#{cask.token}: (#{cask.latest_installed_version}) < #{cask.version}"
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/outdated.rb]: Make this `!=` instead of `<`.

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 8 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: Is there much difference semantically to instead swapping the two lines? Happy to put this first, but I don't understand what the `return false if` adds.

If anything, perhaps pulling this out to a small method e.g. `has_source` would be better from a self documenting perspective.

Happy to go whichever way on this as it seems mostly preferential.

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 9 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: Woops, I added this the other day, but deleted it in a bug fix

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 3 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: This is fixed, but the comment is on the wrong line to pick up the change.

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 8 above)
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: This is mostly just to make it more readable.

`(greedy || !auto_updates && !version == :latest)` is also pretty hard to follow. I'd probably write something like this: 

```ruby
unless greedy
  return false if auto_updates
  return false if version.latest?
end
```

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 11]
```diff
@@ -85,6 +85,20 @@ def installed?
       !versions.empty?
     end
 
+    def latest_installed_version
+      versions.last
+    end
+
+    def has_source?()
+      !version.nil?
+    end
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: I think this was a good idea in theory, but now that I see it, this is even more unclear, since there is no obvious relationship between missing source and version being `nil`.

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 12]
```diff
@@ -85,6 +85,20 @@ def installed?
...
+
+    def outdated?(greedy = false)
+      has_source? &&
+      (greedy || !auto_updates && !version == :latest) &&
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: `!version == :latest` ≠ `version != :latest`

Anyways, you should use `version.latest?`.

File: Library/Homebrew/cask/lib/hbc/cask.rb [hunk 13]
```diff
@@ -85,6 +85,20 @@ def installed?
...
+      version != latest_installed_version.to_s
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: You can remove `to_s` here after https://github.com/Homebrew/brew/pull/2282 is merged.

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 11 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: I agree

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 12 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: Lol yep

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 13 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: I've subscribed to that PR and will update this when it's merged.

**Comment by ilovezfs** [comment]
Is it the case that there is no `brew cask upgrade` commabd? If that's true, it seems like putting the cart before the horse to have a `brew cask outdated` command.

**Comment by joshka** [comment]
> Is it the case that there is no brew cask upgrade command?

No, a brew cask upgrade command does not yet exist. It's been on the drawing board for a very long time. See https://github.com/caskroom/homebrew-cask/issues/29301

> If that's true, it seems like putting the cart before the horse to have a brew cask outdated command.

Not really, the first step to knowing how to upgrade is knowing what to upgrade. I can use `brew cask outdated` to list Casks that require a manual upgrade already, which has some immediate utility despite not being able to do anything other than manually reinstall those. In actual fact `brew cask reinstall $(brew cask outdated)` does exactly what you'd think it should (but doesn't handle the rollback stuff.

See https://github.com/caskroom/homebrew-cask/issues/29301 for much more detail on the spec and rationale for this.

File: Library/Homebrew/cask/lib/hbc/cask.rb (hunk 13 above)
Comment (joshka) [inline Library/Homebrew/cask/lib/hbc/cask.rb]: I removed `to_s` and rebased

**Comment by joshka** [comment]
I''ll add `--verbose` and `--quiet` as well as tests before this is ready to merge

**Comment by joshka** [comment]
TODO: show all versions instead of just the latest version in the outdated list

**Comment by vitorgalvao** [comment]
> TODO: show all versions instead of just the latest version in the outdated list

I wouldn’t argue it that feature did not exist. [As mentioned](https://github.com/caskroom/homebrew-cask/issues/29301#issuecomment-285622306), we only support one version at a time. When you’re moving things to `/Applications`, there’s really no way you actually have two versions installed at the same time.

The goal is even to [stop supporting multiple versions of the same cask](https://github.com/caskroom/homebrew-cask/issues/13201) (item **5**), so I don’t see it as necessary.

**Comment by joshka** [comment]
Closing as https://github.com/Homebrew/brew/pull/2309 is a better version of this.
